modules/qualitycontrol/QualityController.ts

```diff
@@ -231,6 +266,80 @@ export default class QualityController {
+        // Do not attempt run time adjustments if the adaptive mode is disabled.
+        if (!this._enableAdaptiveMode) {
+            return;
+        }
+
+        const { encodeResolution, localTrack, qualityLimitationReason, tpc } = sourceStats;
+        const trackId = localTrack.rtcId.toString();
+
+        if (encodeResolution === tpc.calculateExpectedSendResolution(localTrack)) {
+            if (this._limitedByCpuTimeout) {
+                window.clearTimeout(this._limitedByCpuTimeout);
+                this._limitedByCpuTimeout = undefined;
+            }
+
+            if (qualityLimitationReason === QualityLimitationReason.NONE
+                && this.receiveVideoController.isLastNLimitedByCpu()) {
+                if (!this._lastNRampupTimeout && !this._isLastNRampupBlocked) {
+                    // Ramp up the number of received videos if CPU limitation no longer exists. If the cpu
+                    // limitation returns as a consequence, do not attempt to ramp up again, continue to
+                    // increment the lastN value otherwise until it is equal to the channelLastN value.
+                    this._lastNRampupTimeout = window.setTimeout(() => {
+                        this._lastNRampupTimeout = undefined;
+                        const updatedStats = this._encodeTimeStats.get(trackId);
+                        const latestSourceStats: ISourceStats = updatedStats.get(updatedStats.size() - 1);
+
+                        if (latestSourceStats.qualityLimitationReason === QualityLimitationReason.CPU) {
+                            this._isLastNRampupBlocked = true;
+                        } else {
+                            this._lowerOrRaiseLastN(false /* raise */);
+                        }
+                    }, LAST_N_RAMPUP_TIME);
+                }
+            }
+
+            return;
+        }
+
+        // Do nothing if the limitation reason is bandwidth since the browser will dynamically adapt the outbound
+        // resolution based on available uplink bandwith. Otherwise,
+        // 1. Switch the codec to the lowest complexity one incrementally.
+        // 2. Switch to a lower lastN value, cutting the receive videos by half in every iteration until
+        // MIN_LAST_N value is reached.
+        // 3. Lower the receive resolution of individual streams up to 180p.
+        if (qualityLimitationReason === QualityLimitationReason.CPU
+            || (encodeResolution < tpc.calculateExpectedSendResolution(localTrack)
+                && qualityLimitationReason !== QualityLimitationReason.BANDWIDTH)) {
+
+            if (this._lastNRampupTimeout) {
+                window.clearTimeout(this._lastNRampupTimeout);
+                this._lastNRampupTimeout = undefined;
+                this._isLastNRampupBlocked = true;
+            }
+            const codecSwitched = this._maybeSwitchVideoCodec(trackId);
+
+            if (!codecSwitched && !this._limitedByCpuTimeout) {
```

Comment:
Won't the 2nd part of the confition always be true? We always reset it is set above.

Reply:
Nope thats a different timeout. `_lastNRampupTimeout` is reset above and here we are checking for `_limitedByCpuTimeout`